### PR TITLE
chore: robustly ensure containers are stopped

### PR DESCRIPTION
**Pull request by CaptainCarpensir**
Uses `docker ps` to ensure that the containers are stopped, instead of waiting 1 second and assuming that's enough time to stop them.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/docker/orchestrator/orchestrator.go [hunk 1]
```diff
@@ -219,8 +219,25 @@ func (a *runTaskAction) Do(o *Orchestrator) error {
 		}
 
 		// ensure that containers are fully stopped after o.stopTask finishes blocking
-		// TODO(Aiden): Implement a container ID system or use `docker ps` to ensure containers are stopped
-		time.Sleep(1 * time.Second)
+		for {
```
Comment (dannyrandall) [inline internal/pkg/docker/orchestrator/orchestrator.go]: is there a reason we shouldn't move this into `stopTask()`?

File: internal/pkg/docker/orchestrator/orchestrator.go
Comment (dannyrandall) [inline internal/pkg/docker/orchestrator/orchestrator.go:1]: tests? 🤔 don't want to lose the > 70% coverage 😅

File: internal/pkg/docker/orchestrator/orchestrator.go [hunk 2]
```diff
@@ -219,8 +219,25 @@ func (a *runTaskAction) Do(o *Orchestrator) error {
...
+
+				if running {
+					allCtrsStopped = false
+				}
+			}
+
+			if allCtrsStopped {
+				break
+			}
+		}
```
Comment (dannyrandall) [inline internal/pkg/docker/orchestrator/orchestrator.go]: seems like we should stop the whole loop if `ctx` is cancelled (like when someone calls Stop()) - something like:
```
for {
  select {
  case <-ctx.Done():
    return ctx.Err()
  case <-time.After(100 * time.Millisecond):
    // the above logic
  }
}
```

File: internal/pkg/docker/orchestrator/orchestrator.go [hunk 3]
```diff
@@ -411,8 +407,33 @@ func (o *Orchestrator) stopTask(ctx context.Context, task Task) error {
...
-	return errors.Join(errs...)
+	// ensure that containers are fully stopped before stopTask finishes blocking
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(100 * time.Millisecond):
+			allCtrsStopped := true
+			for ctrName := range o.curTask.Containers {
+				running, err := o.docker.IsContainerRunning(ctx, o.containerID(ctrName))
```
Comment (dannyrandall) [inline internal/pkg/docker/orchestrator/orchestrator.go]: what do you think about moving this into the goroutine above we create for each container? basically, right after each `docker.Stop()` call, we don't send anything to `errCh` until it's stopped 🤔 then we can get rid of the `allCtrsStopped` logic and just assume no error returned == success, and additionally get an error for _each_ container that we're unable to get the status for

Comment (CaptainCarpensir) [reply internal/pkg/docker/orchestrator/orchestrator.go]: Really like this! I don't enjoy the `allCtrsStopped` logic but I couldn't think of a way around it, this should be more accurate.

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/docker/orchestrator/orchestrator.go [hunk 4]
```diff
@@ -399,8 +395,29 @@ func (o *Orchestrator) stopTask(ctx context.Context, task Task) error {
...
+					errCh <- fmt.Errorf("polling container %q for removal: %w", name, err)
+					return
+				}
+
+				if running {
+					select {
+					case <-time.After(1 * time.Second):
+						continue
+					case <-ctx.Done():
+						errCh <- ctx.Err()
```
Comment (dannyrandall) [inline internal/pkg/docker/orchestrator/orchestrator.go]: maybe should add `name` context to the error here?

**Review by dannyrandall — APPROVED** [review]
looks great! just a small nit 😊

**Pull request merged**